pkg/storage/tsdb/caching_bucket.go

```diff
@@ -36,47 +62,133 @@ func (cfg *ChunksCacheConfig) RegisterFlagsWithPrefix(f *flag.FlagSet, prefix st
+	f.StringVar(&cfg.Backend, prefix+"backend", "", fmt.Sprintf("Backend for metadata cache, if not empty. Supported values: %s.", CacheBackendMemcached))
+
+	cfg.Memcached.RegisterFlagsWithPrefix(f, prefix+"memcached.")
+
+	f.DurationVar(&cfg.TenantsListTTL, prefix+"tenants-list-ttl", 15*time.Minute, "How long to cache list of tenants in the bucket.")
+	f.DurationVar(&cfg.TenantBlocksListTTL, prefix+"tenant-blocks-list-ttl", 15*time.Minute, "How long to cache list of blocks for each tenant.")
+	f.DurationVar(&cfg.ChunksListTTL, prefix+"chunks-list-ttl", 24*time.Hour, "How long to cache list of chunks for a block.")
+	f.DurationVar(&cfg.MetafileExistsTTL, prefix+"metafile-exists-ttl", 2*time.Hour, "How long to cache information that block metafile exists.")
+	f.DurationVar(&cfg.MetafileDoesntExistTTL, prefix+"metafile-doesnt-exist-ttl", 15*time.Minute, "How long to cache information that block metafile doesn't exist.")
+	f.DurationVar(&cfg.MetafileContentTTL, prefix+"metafile-content-ttl", 24*time.Hour, "How long to cache content of the metafile.")
+	f.IntVar(&cfg.MetafileMaxSize, prefix+"metafile-max-size", 1*1024*1024, "Maximum size of metafile content to cache.")
+}
+
+func (cfg *MetadataCacheConfig) Validate() error {
+	return cfg.CacheBackend.Validate()
+}
+
+func CreateCachingBucket(chunksConfig ChunksCacheConfig, metadataConfig MetadataCacheConfig, bkt objstore.Bucket, logger log.Logger, reg prometheus.Registerer) (objstore.Bucket, error) {
+	cfg := storecache.NewCachingBucketConfig()
+	cachingConfigured := false
+
+	chunksCache, err := createCache("chunks-cache", chunksConfig.Backend, chunksConfig.Memcached, logger, reg)
+	if err != nil {
+		return nil, errors.Wrapf(err, "chunks-cache")
+	}
+	if chunksCache != nil {
+		cachingConfigured = true
+		cfg.CacheGetRange("chunks", chunksCache, isTSDBChunkFile, chunksConfig.SubrangeSize, chunksConfig.ObjectSizeTTL, chunksConfig.SubrangeTTL, chunksConfig.MaxGetRangeRequests)
 	}
 
-	if cfg.Backend == string(storecache.MemcachedBucketCacheProvider) {
-		if err := cfg.Memcached.Validate(); err != nil {
-			return err
-		}
+	metadataCache, err := createCache("metadata-cache", metadataConfig.Backend, metadataConfig.Memcached, logger, reg)
+	if err != nil {
+		return nil, errors.Wrapf(err, "metadata-cache")
 	}
+	if metadataCache != nil {
+		cachingConfigured = true
 
-	return nil
-}
+		cfg.CacheExists("metafile", metadataCache, isMetaFile, metadataConfig.MetafileExistsTTL, metadataConfig.MetafileDoesntExistTTL)
+		cfg.CacheGet("metafile", metadataCache, isMetaFile, metadataConfig.MetafileMaxSize, metadataConfig.MetafileContentTTL, metadataConfig.MetafileExistsTTL, metadataConfig.MetafileDoesntExistTTL)
 
-func CreateCachingBucket(chunksConfig ChunksCacheConfig, bkt objstore.Bucket, logger log.Logger, reg prometheus.Registerer) (objstore.Bucket, error) {
-	var chunksCache cache.Cache
+		cfg.CacheIter("tenants-iter", metadataCache, isTenantsDir, metadataConfig.TenantsListTTL, storecache.JSONIterCodec{})
```

Comment:
Why not compress this too with snappy? If you have 10K tenants, we could save some bytes. Given it's snappy, if you only have 1 tenant there's no bytes saving but also the impact on performances would be not noticeable.

Reply:
We can do that. While I don't expect compression ratio to be great for root directory, as you note, performance impact is minimal.